### PR TITLE
Generate r2 score figures and sc crops

### PR DESCRIPTION
**Pull request by jenna-tomkinson**
# Generate r2 score figures and sc crops

In this PR, R2 scores are are calculated and figures are generated. As well, min/max single-cell crops are generated but not saved to the repo.

### CODE REVIEW

File: 1.regression_modelling/scripts/1.train_models.py [hunk 1]
```diff
@@ -104,6 +105,23 @@
...
+
+# Organize the column names into a dictionary
+features_dict = {
+    "nucleus_features": nucleus_features,
+    "a647_features": a647_features,
+    "gold_features": gold_features,
+}
+
+# Define the path to save the YAML file
+yaml_file_path = pathlib.Path("./features_dict.yml")
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/1.train_models.py]: Consider resolving this path?

File: 1.regression_modelling/scripts/2.r2_figure.py [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+from sklearn.metrics import r2_score
+
+
+# ## Set paths
+
+# In[2]:
+
+
+# Output for figures
+figures_dir = pathlib.Path("./figures")
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/2.r2_figure.py]: same here

File: 1.regression_modelling/scripts/2.r2_figure.py [hunk 3]
```diff
@@ -0,0 +1,657 @@
...
+figures_dir.mkdir(parents=True, exist_ok=True)
+
+
+# ## Define features per stain to use for filtering columns
+
+# In[3]:
+
+
+# Define the path to the YAML file with all the features per stain
+yaml_file_path = pathlib.Path("./features_dict.yml")
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/2.r2_figure.py]: and here

File: 1.regression_modelling/scripts/2.r2_figure.py [hunk 4]
```diff
@@ -0,0 +1,657 @@
...
+
+# Access the lists of features
+nucleus_features = features_dict['nucleus_features']
+a647_features = features_dict['a647_features']
+gold_features = features_dict['gold_features']
+
+# Now you can use these lists for further operations
+print(f"Nucleus Features: {len(nucleus_features)}")
+print(f"A647 Features: {len(a647_features)}")
+print(f"Gold Features: {len(gold_features)}")
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/2.r2_figure.py]: I like this import and feature organization style! No data wrangling every time. 

Are these feature selected features? Might be useful to include in the yaml as a comment?

File: 1.regression_modelling/scripts/2.r2_figure.py [hunk 5]
```diff
@@ -0,0 +1,657 @@
...
+    plt.legend(title='Measurement', bbox_to_anchor=(1.05, 1), loc='upper left')
+
+    # Improve layout
+    plt.tight_layout()
+
+    # Save
+    plt.savefig(f"./{figures_dir}/shuffled_versus_final_r2.png", dpi=600)
+
+    # Show plot
+    plt.show()
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/2.r2_figure.py]: I would recommend adding type hints for this function and possible a small doc string? Also consider adding a arg of the save path to the function so that global args are not required

File: 1.regression_modelling/scripts/2.r2_figure.py [hunk 6]
```diff
@@ -0,0 +1,657 @@
...
+
+# ## Split features to get an R2 score per stain and plot R2 results
+
+# ### Function to split features based on stain to have one R2 score per stain
+
+# In[11]:
+
+
+def process_subset(df, split_type, model_type):
+    # Create a copy of df_subset to avoid modifying the original slice
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/2.r2_figure.py]: same as above for this function

File: 1.regression_modelling/scripts/2.r2_figure.py [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+
+
+# Set the figure size
+plt.figure(figsize=(12, 8))
+
+# Create the vertical box plot
+sns.boxplot(
+    data=filtered_training_df,
+    x='Metadata_Condition',  # Metadata_condition on x-axis
+    y='Nuclei_RadialDistribution_FracAtD_GOLD_4of4',  # Nuclei_RadialDistribution_FracAtD_GOLD_4of4 on y-axis
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/2.r2_figure.py:563]: any way to make these labels more dynamic?

File: 1.regression_modelling/scripts/2.r2_figure.py [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+
+# Show plot
+plt.show()
+
+
+# In[ ]:
+
+
+
+
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/2.r2_figure.py]: consider removing

File: 1.regression_modelling/scripts/3.find_sc_crops.py [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+                "location_center_y": row["Metadata_Nuclei_Location_Center_Y"],
+            }
+    return sc_dict
+
+
+# In[3]:
+
+
+# Function for generating and saving single-cell crops per channel as PNGs
+def generate_sc_crops(
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/3.find_sc_crops.py:54]: Seeing how these functions build off eachother. Consider maybe making these class methods in a class. No pressure if not.

File: 1.regression_modelling/scripts/3.find_sc_crops.py [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        output_img_dir (pathlib.Path): Main directory to save each image set single-cell crops
+        crop_size (int): Size of the box in pixels (example: setting crop_size as 250 will make a 250x250 pixel crop around the single-cell center coordinates)
+    """
+    for key, info in sc_dict.items():
+        # Initialize a list to store file paths for every image set
+        file_paths = []
+
+        # Create file paths with well, site, and channel
+        for i in range(3):  # Update the range to start from 0 and end at 2
+            filename = f"{images_dir}/{info['plate']}/{info['plate']}_{info['well']}_{info['site']}_CH{i}_Z09_illumcorrect.tiff"
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/3.find_sc_crops.py:75]: This is organized but stressful haha. Good job managing chaos into organization.

File: 1.regression_modelling/scripts/3.find_sc_crops.py [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+# In[5]:
+
+
+model_dir = pathlib.Path('./models')  # Path to the folder with models
+
+# paths to the data splits
+training_data_path = pathlib.Path('./data/training_data.parquet')
+
+# Load your datasets
+training_df = pd.read_parquet(training_data_path)
```
Comment (MikeLippincott) [inline 1.regression_modelling/scripts/3.find_sc_crops.py]: consider resolving these

File: 1.regression_modelling/figures/holdout_a647_versus_gold_r2.png
Comment (MikeLippincott) [inline 1.regression_modelling/figures/holdout_a647_versus_gold_r2.png:1]: Nitpicks on the basis of being used for a presentation not a publication:
1. Is the Y axis the predicted stain? If yes, consider adjusting the axis label.
2. The text all around is a bit small
3. I would prefer (my preference here) to make the x axis go all the way 1.0 instead of 0.75
4. It would be nice if the dots were in the same bounds or margins as the bars. This is my thought, feel free to ignore
5. It would be nice to see the other data splits in a separate (or same) plot. I thought I saw those in your code?

File: 1.regression_modelling/figures/holdout_versus_testing_scatter_r2.png
Comment (MikeLippincott) [inline 1.regression_modelling/figures/holdout_versus_testing_scatter_r2.png:1]: This is cool that certain features can be predicted high in both stains. For the low performing features, were these influenced more by certain treatments? Is this at the aggregate level?

File: 1.regression_modelling/figures/shuffled_versus_final_r2.png
Comment (MikeLippincott) [inline 1.regression_modelling/figures/shuffled_versus_final_r2.png:1]: Nice, this is sick! Same nitpicks as above.

File: 1.regression_modelling/figures/holdout_a647_versus_gold_r2.png
Comment (MikeLippincott) [reply 1.regression_modelling/figures/holdout_a647_versus_gold_r2.png:1]: Also was there a shuffled model evaluated for the holdout data?

File: 1.regression_modelling/figures/holdout_versus_testing_scatter_r2.png
Comment (MikeLippincott) [reply 1.regression_modelling/figures/holdout_versus_testing_scatter_r2.png:1]: Shuffle models?

File: 1.regression_modelling/figures/shuffled_versus_final_r2.png
Comment (MikeLippincott) [reply 1.regression_modelling/figures/shuffled_versus_final_r2.png:1]: Other data splits too?

File: 1.regression_modelling/figures/training_versus_testing_scatter_r2.png
Comment (MikeLippincott) [inline 1.regression_modelling/figures/training_versus_testing_scatter_r2.png:1]: I think I have asked this question on a past PR: What are GOLD and A647 staining? Are these fluorophores tagged to anything or is the tagged protein variable across samples?

**Review by MikeLippincott — APPROVED** [review]
This LGTM. Not sure if all of the figures are needed now if you are going to remake them in R later, but it seems that there are multiple other figures that would add value to this analysis. Please consider adding them either now or in a future PR!

File: 1.regression_modelling/scripts/1.train_models.py (hunk 1 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/1.train_models.py]: Added!

File: 1.regression_modelling/scripts/2.r2_figure.py (hunk 2 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/2.r2_figure.py]: Added!

File: 1.regression_modelling/scripts/2.r2_figure.py (hunk 3 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/2.r2_figure.py]: Added!

File: 1.regression_modelling/scripts/2.r2_figure.py (hunk 4 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/2.r2_figure.py]: Thanks! :smile: These are the normalized features that are exported as a yml file in the previous notebook. I will make a code comment to mention it here.

File: 1.regression_modelling/scripts/2.r2_figure.py (hunk 5 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/2.r2_figure.py]: Ahh I forgot to add that. I will add it in now thanks!

File: 1.regression_modelling/scripts/2.r2_figure.py (hunk 8 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/2.r2_figure.py]: Thanks for catching that! Removed!

File: 1.regression_modelling/scripts/2.r2_figure.py (hunk 6 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/2.r2_figure.py]: Added! Thanks for catching!

File: 1.regression_modelling/scripts/3.find_sc_crops.py (hunk 9 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/3.find_sc_crops.py:54]: I think for now I will keep it as two functions for now. I will consider class in the future if I plan to build more into generating these crops. I appreciate the suggestion!

File: 1.regression_modelling/scripts/3.find_sc_crops.py (hunk 10 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/3.find_sc_crops.py:75]: Haha thank you! 😄

File: 1.regression_modelling/scripts/3.find_sc_crops.py (hunk 11 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/3.find_sc_crops.py]: Added!

File: 1.regression_modelling/scripts/2.r2_figure.py (hunk 7 above)
Comment (jenna-tomkinson) [reply 1.regression_modelling/scripts/2.r2_figure.py:563]: I think for now I will keep it hardcoded since we likely won't be rerunning these single-models and find different top features.

File: 1.regression_modelling/figures/holdout_a647_versus_gold_r2.png
Comment (jenna-tomkinson) [reply 1.regression_modelling/figures/holdout_a647_versus_gold_r2.png:1]: 1. The Y-axis is what stain the feature being predicted comes from, so I think this title is sufficient for now.
2. Agreed, but it was sufficient for the presentation, so I will keep it for now.
3. Understandable, I think it not going to 1 is okay since there is really only one model that stands out and that is the main point.
4. This is due to jitter to avoid too much overlap of the dots. I could reduce it but I think it is sufficient for this figure.
5. These other splits are visualized in the scatterplot instead as it gives more information on which stain has better performance.
6. Yes there is a shuffled model but in the training data, it shows that all models perform around 0 so I have a different figure to show that the shuffled performs worse.

File: 1.regression_modelling/figures/holdout_versus_testing_scatter_r2.png
Comment (jenna-tomkinson) [reply 1.regression_modelling/figures/holdout_versus_testing_scatter_r2.png:1]: I do not believe treatment made much of a contribution in the low performing features. Actually in this case, most of the features at 0 are way below 0 for the holdout data. 

This is at the single-cell level. As well, for the shuffled models, these performed very poorly and introduced too much noise in the plot that I wanted to just focus on the difference in performance between testing and holdout which are two different cell lines.

File: 1.regression_modelling/figures/shuffled_versus_final_r2.png
Comment (jenna-tomkinson) [reply 1.regression_modelling/figures/shuffled_versus_final_r2.png:1]: Yes there is but for the presentation I wanted to just focus on the testing data + when I made it with the training data the plot looked almost the exact same so I decided to just only include the testing as a way to demonstrate the difference between final and shuffled.

File: 1.regression_modelling/figures/training_versus_testing_scatter_r2.png
Comment (jenna-tomkinson) [reply 1.regression_modelling/figures/training_versus_testing_scatter_r2.png:1]: Yes you have lol These stains are for the protein markers related to nuclear speckles. Here is what I have in the README for the profiling repo:

```
We have 3 channels in this assay:

    DAPI - nucleus stain
    A647 - SON protein marker for nuclear speckles
    Cy5/GOLD - SRRM2 protein marker for nuclear speckles
```

**Comment by jenna-tomkinson** [comment]
@MikeLippincott Thank you for the review! I am merging now!
